src/pygeosx/pysolver/PySolver.cpp

```diff
@@ -0,0 +1,303 @@
+    {
+      Py_DECREF( self );
+      return nullptr;
+    }
+
+  }
+
+  return (PyObject *)self;
+}
+
+
+
+static int PySolver_init( PySolver *self, PyObject *args, PyObject *kwds )
+{
+  GEOSX_UNUSED_VAR( kwds );
+
+  PYTHON_ERROR_IF( !g_alreadyInitialized, PyExc_RuntimeError, "GEOSX must be initialized first. Call pygeosx.initialize().", -1 );
+
+  geosx::SolverBase *group=nullptr;
+
+  PyObject * unicodePath;
+  if( !PyArg_ParseTuple( args, "U", &unicodePath ) )
+  {
+    return -1;
+  }
+
+  LvArray::python::PyObjectRef<> asciiPath { PyUnicode_AsASCIIString( unicodePath ) };
+  if( asciiPath == nullptr )
+  {
+    return -1;
+  }
+
+  char const * const path = PyBytes_AsString( asciiPath );
+  if( path == nullptr )
+  {
+    return -1;
+  }
+
+  geosx::EventManager & eventManager = self->pb_manager->getEventManager();
+
+  eventManager.forSubGroups< geosx::EventBase >( [&]( geosx::EventBase & subEvent )
+  {
+    subEvent.getTargetReferences();
+  } );
+
+  geosx::EventBase * subEvent = nullptr;
+  for( int currentSubEvent = 0; currentSubEvent<eventManager.numSubGroups(); ++currentSubEvent )
+  {
+
+    subEvent = static_cast< geosx::EventBase * >( eventManager.getSubGroups()[currentSubEvent]);
+    if( subEvent->getEventName() == path )
+    {
+      break;
+    }
+    else
+    {
+      subEvent = nullptr;
+    }
+  }
+  PYTHON_ERROR_IF( subEvent == nullptr, PyExc_RuntimeError, "Target not found", -1 );
+
+  group = static_cast< geosx::SolverBase * >(subEvent->getEventTarget());
+  self->group = group;
+
+  return 0;
+}
+
+
+static PyObject * PySolver_repr( PyObject * const obj ) noexcept
+{
+  PySolver const * const pySolver = LvArray::python::convert< PySolver >( obj, getPySolverType() );
+  if( pySolver == nullptr )
+  {
+    return nullptr;
+  }
+
+  VERIFY_INITIALIZED( pySolver );
+
+  string const path = pySolver->group->getPath();
+  string const type = LvArray::system::demangle( typeid( *(pySolver->group) ).name() );
+  string const repr = path + " ( " + type + " )";
+
+  return PyUnicode_FromString( repr.c_str() );
+}
+
+
+
+static PyObject * postProcessInput( PySolver * self, PyObject * args )
+{
+  VERIFY_NON_NULL_SELF( self );
+  VERIFY_INITIALIZED( self );
+  GEOSX_UNUSED_VAR( args );
+
+  self->group->postProcessInput();
+
+  Py_RETURN_NONE;
+}
+
+
+static PyObject * initPostInitialConditions( PySolver * self, PyObject * args )
+{
+  VERIFY_NON_NULL_SELF( self );
+  VERIFY_INITIALIZED( self );
+  GEOSX_UNUSED_VAR( args );
+
+  self->group->initializePostInitialConditionsPreSubGroups();
+
+  Py_RETURN_NONE;
+}
+
+
+static PyObject * explicitStep( PySolver * self, PyObject * args )
+{
+  VERIFY_NON_NULL_SELF( self );
+  VERIFY_INITIALIZED( self );
+
+  double time;
+  double dt;
+  if( !PyArg_ParseTuple( args, "dd", &time, &dt ) )
+  {
+    return nullptr;
+  }
+
+  geosx::DomainPartition & domain = self->pb_manager->getDomainPartition();
+
+  self->group->explicitStep( time, dt, 0, domain );
```

Comment:
What happens if you've got an implicit solver but you call this function?

Reply:
```
***** ERROR
***** LOCATION: /home/julien/codes/GEOSX/src/coreComponents/physicsSolvers/SolverBase.cpp:647
***** Controlling expression (should be false): true
***** Rank 0: SolverBase::ImplicitStepSetup called!. Should be overridden.
```
Is there a way to `catch` a wrong call method ?

Reply:
Is this raised as a Python exception or does it kill the program? The first is fine, but the second isn't (just change it from a `GEOSX_ERROR_IF` to `GEOSX_THROW_IF`, and then wrap it in a `try-except` block. But what about if you call it on a solver that can do both, like the `SolidMechanics`?

Reply:
Done.
For the case where a `Solver` can do both, could we add flags in the solvers` that would tell if it's using `linear` or `non linear` and `explicit` or `implicit` depending on what is defined in the XML ? Then we would rely on that on `PySolver` side

Reply:
What's the reason we're not simply exposing `solverStep` as the main "driver" function for a solver? GEOSX solvers already implement switching logic for different step types, so why shift the burden on user? (I'm a little out of context on this PR, so apologies if the answer is obvious).

Reply:
Actually I think you're absolutely right @klevzoff. It should be sufficient to just expose `solverStep`

Reply:
10/10 solution.